### ISSUE TITLE
Errors while downloading mcumgr

### ISSUE THREAD

**Issue by matwiszynmarek** [apache/mynewt-mcumgr-cli#8]
Hello everyone.
I'm trying to get mcumgr using a go command but I get such errors :+1: 

```sh
$ go get github.com/apache/mynewt-mcumgr-cli/mcumgr
# mynewt.apache.org/newtmgr/newtmgr/cli
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:102:22: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:144:16: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:168:21: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:191:23: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:255:22: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:267:16: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:321:17: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:375:19: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:417:24: undefined: ishell.Context
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:444:7: shell.AddCmd undefined (type *ishell.Shell has no field or method AddCmd)
go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:444:7: too many errors
```

I've never had those errors before. Any suggestion where should I find a bug?

**Comment by lawrencek52** [comment]
This is my first time installing go, and attempting to install mcumgr and I am seeing exactly the same errors. What piece are we missing? I am using "go version go1.12.7 linux/amd64".

**Comment by ccollins476ad** [comment]
Sorry for missing this issue when it was first posted.

Unfortunately I am not able to reproduce this.  Can either of you please tell me which directory you are running the `go get` command from?  Just be sure you are not in another go project's directory when you run it.

If that is not the problem, you might try setting this environment variable first:
```
export GO111MODULE=on
```

**Comment by lawrencek52** [comment]
```
lawrence@Fred:~$ mkdir foo
lawrence@Fred:~$ cd foo
lawrence@Fred:~/foo$ export GO111MODULE=on
lawrence@Fred:~/foo$ go get github.com/apache/mynewt-mcumgr-cli/mcumgr
# mynewt.apache.org/newtmgr/newtmgr/cli
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:134:18: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:164:16: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:193:21: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:249:23: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:304:22: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:326:16: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:356:17: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:386:19: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:409:24: undefined: ishell.Context
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:439:7: shell.AddCmd undefined (type *ishell.Shell has no field or method AddCmd)
../.go/src/mynewt.apache.org/newtmgr/newtmgr/cli/interactive.go:439:7: too many errors
lawrence@Fred:~/foo$ go version
go version go1.10.4 linux/amd64
```

**Comment by lawrencek52** [comment]
I changed how I installed go from
`sudo snap install go --classic`
to
`sudo apt install golang`
Hence the change in versions
Unfortunately the same problem either way.

**Comment by ccollins476ad** [comment]
@lawrencek52 would you mind posting the `go get` output when using go 1.12 (if you have it handy)?

**Comment by SIGSTACKFAULT** [comment]
This patch **to newtmgr** seems to fix it. It's probably a bad workaround, though.
```
diff --git a/newtmgr/cli/interactive.go b/newtmgr/cli/interactive.go
index e083662..8c7909a 100644
--- a/newtmgr/cli/interactive.go
+++ b/newtmgr/cli/interactive.go
@@ -29,7 +29,7 @@ import (
 	"github.com/runtimeco/go-coap"
 	"github.com/spf13/cast"
 	"github.com/spf13/cobra"
-	"gopkg.in/abiosoft/ishell.v1"
+	"github.com/abiosoft/ishell"
 	"mynewt.apache.org/newtmgr/newtmgr/nmutil"
 	"mynewt.apache.org/newtmgr/nmxact/nmcoap"
 	"mynewt.apache.org/newtmgr/nmxact/nmxutil"
```

**Comment by utzig** [comment]
@SIGSTACKFAULT Looks good, feel free to send a PR!

**Comment by SIGSTACKFAULT** [comment]
It's not like this line has changed recently. excerpt from `git blame interactive.go`:
```
346002cd (Christopher Collins 2018-02-20 13:32:56 -0800  32)    "gopkg.in/abiosoft/ishell.v1"
```
_Fascinating._

**Comment by lawrencek52** [comment]
> @lawrencek52 would you mind posting the `go get` output when using go 1.12 (if you have it handy)?

Other than the version reported by 'go version', the error messages are identical with both versions of go. 

The @SIGSTACKFAULT solution works. Thanks

**Comment by utzig** [comment]
> Fascinating.

Yeah, `go get` also fails for me, although I have a local clone of `mynewt-newtmgr` and doing `go build` there just works!

**Comment by ccollins476ad** [comment]
@lawrencek52 I expected to see some "finding..." lines when you use go 1.12 and `GO111MODULE` enabled, e.g.
```
[ccollins@ccollins:~]$ go version
go version go1.12.7 darwin/amd64
[ccollins@ccollins:~]$ go get github.com/apache/mynewt-mcumgr-cli/mcumgr
go: finding github.com/apache/mynewt-mcumgr-cli/mcumgr latest
go: finding github.com/apache/mynewt-mcumgr-cli latest
```

But I take your word that the output is the same.  I'll keep looking.

**Comment by ccollins476ad** [comment]
Maybe we just need to update newtmgr.  From https://github.com/abiosoft/ishell:
> The current master is not backward compatible with older version. Kindly change your import path to gopkg.in/abiosoft/ishell.v1.

**Comment by utzig** [comment]
Oops, I was missing `GO111MODULE=on` here, setting it actually works.

```
$ go get -v github.com/apache/mynewt-mcumgr-cli/mcumgr
go: finding github.com/apache/mynewt-mcumgr-cli/mcumgr latest
go: finding github.com/apache/mynewt-mcumgr-cli latest
go: downloading github.com/apache/mynewt-mcumgr-cli v0.0.0-20190801172116-982b276df3a2
go: extracting github.com/apache/mynewt-mcumgr-cli v0.0.0-20190801172116-982b276df3a2
Fetching https://mynewt.apache.org/newtmgr?go-get=1
Fetching https://mynewt.apache.org/newt?go-get=1
Parsing meta tags from https://mynewt.apache.org/newtmgr?go-get=1 (status code 200)
Parsing meta tags from https://mynewt.apache.org/newt?go-get=1 (status code 200)
get "mynewt.apache.org/newt": found meta tag get.metaImport{Prefix:"mynewt.apache.org/newt", VCS:"git", RepoRoot:"https://github.com/apache/mynewt-newt.git"} at https://mynewt.apache.org/newt?go-get=1
get "mynewt.apache.org/newtmgr": found meta tag get.metaImport{Prefix:"mynewt.apache.org/newtmgr", VCS:"git", RepoRoot:"https://github.com/apache/mynewt-newtmgr.git"} at https://mynewt.apache.org/newtmgr?go-get=1
go: finding mynewt.apache.org/newt v0.0.0-20190723172213-9f7829a76fef
go: finding mynewt.apache.org/newtmgr v0.0.0-20190801165443-4ba7e39aca9f
Fetching https://golang.org/x/crypto?go-get=1
go: finding github.com/spf13/cast v1.3.0
go: finding github.com/ugorji/go/codec v1.1.7
go: finding github.com/ugorji/go v1.1.7
go: finding github.com/sirupsen/logrus v1.4.2
go: finding github.com/spf13/cobra v0.0.5
go: finding github.com/NickBall/go-aes-key-wrap v0.0.0-20170929221519-1c3aa3e4dfc5
go: finding github.com/spf13/pflag v1.0.3
go: finding github.com/apache/mynewt-artifact v0.0.3
Parsing meta tags from https://golang.org/x/crypto?go-get=1 (status code 200)
get "golang.org/x/crypto": found meta tag get.metaImport{Prefix:"golang.org/x/crypto", VCS:"git", RepoRoot:"https://go.googlesource.com/crypto"} at https://golang.org/x/crypto?go-get=1
go: finding golang.org/x/crypto v0.0.0-20190621222207-cc06ce4a13d4
go: finding github.com/stretchr/testify v1.2.2
go: finding github.com/pmezard/go-difflib v1.0.0
go: finding github.com/davecgh/go-spew v1.1.1
go: finding golang.org/x/crypto v0.0.0-20190618222545-ea8f1a30c443
go: finding github.com/stretchr/testify v1.3.0
go: finding github.com/cpuguy83/go-md2man v1.0.10
go: finding github.com/pkg/errors v0.8.1
go: finding github.com/spf13/viper v1.3.2
go: finding github.com/inconshreveable/mousetrap v1.0.0
go: finding github.com/stretchr/objx v0.1.1
Fetching https://golang.org/x/sys?go-get=1
Parsing meta tags from https://golang.org/x/sys?go-get=1 (status code 200)
get "golang.org/x/sys": found meta tag get.metaImport{Prefix:"golang.org/x/sys", VCS:"git", RepoRoot:"https://go.googlesource.com/sys"} at https://golang.org/x/sys?go-get=1
go: finding github.com/konsorten/go-windows-terminal-sequences v1.0.1
go: finding golang.org/x/sys v0.0.0-20190422165155-953cdadca894
go: finding github.com/fsnotify/fsnotify v1.4.7
go: finding github.com/spf13/afero v1.1.2
Fetching https://gopkg.in/yaml.v2?go-get=1
go: finding github.com/hashicorp/hcl v1.0.0
go: finding github.com/coreos/go-etcd v2.0.0+incompatible
Parsing meta tags from https://gopkg.in/yaml.v2?go-get=1 (status code 200)
get "gopkg.in/yaml.v2": found meta tag get.metaImport{Prefix:"gopkg.in/yaml.v2", VCS:"git", RepoRoot:"https://gopkg.in/yaml.v2"} at https://gopkg.in/yaml.v2?go-get=1
go: finding gopkg.in/yaml.v2 v2.2.2
go: finding github.com/stretchr/objx v0.1.0
go: finding github.com/coreos/etcd v3.3.10+incompatible
go: finding github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6
go: finding github.com/pelletier/go-toml v1.2.0
go: finding github.com/ugorji/go/codec v0.0.0-20181204163529-d75b2dcb6bc8
go: finding github.com/coreos/go-semver v0.2.0
go: finding github.com/russross/blackfriday v1.5.2
go: finding github.com/BurntSushi/toml v0.3.1
go: finding github.com/spf13/jwalterweatherman v1.0.0
go: finding github.com/davecgh/go-spew v1.1.0
Fetching https://golang.org/x/net?go-get=1
go: finding github.com/mitchellh/go-homedir v1.1.0
Parsing meta tags from https://golang.org/x/net?go-get=1 (status code 200)
get "golang.org/x/net": found meta tag get.metaImport{Prefix:"golang.org/x/net", VCS:"git", RepoRoot:"https://go.googlesource.com/net"} at https://golang.org/x/net?go-get=1
go: finding golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3
go: finding github.com/mitchellh/mapstructure v1.1.2
go: finding golang.org/x/sys v0.0.0-20181205085412-a5c9d58dba9a
go: finding github.com/xordataexchange/crypt v0.0.3-0.20170626215501-b2862e3d0a77
go: finding golang.org/x/sys v0.0.0-20190412213103-97732733099d
go: finding golang.org/x/crypto v0.0.0-20181203042331-505ab145d0a9
Fetching https://golang.org/x/text?go-get=1
Parsing meta tags from https://golang.org/x/text?go-get=1 (status code 200)
get "golang.org/x/text": found meta tag get.metaImport{Prefix:"golang.org/x/text", VCS:"git", RepoRoot:"https://go.googlesource.com/text"} at https://golang.org/x/text?go-get=1
go: finding golang.org/x/text v0.3.0
go: finding github.com/magiconair/properties v1.8.0
Fetching https://gopkg.in/check.v1?go-get=1
Parsing meta tags from https://gopkg.in/check.v1?go-get=1 (status code 200)
get "gopkg.in/check.v1": found meta tag get.metaImport{Prefix:"gopkg.in/check.v1", VCS:"git", RepoRoot:"https://gopkg.in/check.v1"} at https://gopkg.in/check.v1?go-get=1
go: finding gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
go: finding golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2
go: finding golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a
Fetching https://gopkg.in/abiosoft/ishell.v1?go-get=1
go: finding github.com/mitchellh/go-homedir v0.0.0-20161203194507-b8bc1bf76747
go: finding github.com/fatih/structs v1.0.0
go: finding github.com/cheggaaa/pb v1.0.18
go: finding github.com/mgutz/logxi v0.0.0-20161027140823-aebf8a7d67ab
go: finding github.com/abiosoft/ishell v2.0.0+incompatible
go: finding github.com/chzyer/test v0.0.0-20180213035817-a1ea475d72b1
go: finding github.com/mattn/go-runewidth v0.0.2
go: finding github.com/flynn-archive/go-shlex v0.0.0-20150515145356-3f9db97f8568
go: finding github.com/abiosoft/readline v0.0.0-20180607040430-155bce2042db
Parsing meta tags from https://gopkg.in/abiosoft/ishell.v1?go-get=1 (status code 200)
get "gopkg.in/abiosoft/ishell.v1": found meta tag get.metaImport{Prefix:"gopkg.in/abiosoft/ishell.v1", VCS:"git", RepoRoot:"https://gopkg.in/abiosoft/ishell.v1"} at https://gopkg.in/abiosoft/ishell.v1?go-get=1
go: finding gopkg.in/abiosoft/ishell.v1 v1.0.0-20171224170712-50251d04cb42
go: finding github.com/joaojeronimo/go-crc16 v0.0.0-20140729130949-59bd0194935e
go: finding github.com/fatih/color v1.5.0
go: finding github.com/go-ble/ble v0.0.0-20181002102605-e78417b510a3
Fetching https://gopkg.in/cheggaaa/pb.v1?go-get=1
go: finding github.com/runtimeco/go-coap v0.0.0-20180109101419-47c23e2a63ea
go: finding github.com/chzyer/logex v1.1.10
Parsing meta tags from https://gopkg.in/cheggaaa/pb.v1?go-get=1 (status code 200)
get "gopkg.in/cheggaaa/pb.v1": found meta tag get.metaImport{Prefix:"gopkg.in/cheggaaa/pb.v1", VCS:"git", RepoRoot:"https://gopkg.in/cheggaaa/pb.v1"} at https://gopkg.in/cheggaaa/pb.v1?go-get=1
go: finding gopkg.in/cheggaaa/pb.v1 v1.0.28
go: finding github.com/tarm/serial v0.0.0-20180830185346-98f6abe2eb07
go: finding github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
go: finding github.com/pkg/errors v0.8.0
go: finding golang.org/x/net v0.0.0-20170828231752-66aacef3dd8a
go: finding github.com/mattn/go-colorable v0.0.9
go: finding github.com/chzyer/readline v0.0.0-20171208011716-f6d7a1f6fbf3
go: finding github.com/raff/goble v0.0.0-20170814221208-b12b34f940c4
go: finding github.com/mattn/go-isatty v0.0.2
go: finding github.com/sirupsen/logrus v1.4.1
go: finding github.com/spf13/cast v1.2.0
go: finding mynewt.apache.org/newt v0.0.0-20190515182908-60fc2df06bfc
go: finding golang.org/x/crypto v0.0.0-20180621125126-a49355c7e3f8
go: finding github.com/sirupsen/logrus v1.0.5
go: finding golang.org/x/sys v0.0.0-20180707002001-3c6ecd8f22c6
go: finding github.com/spf13/pflag v1.0.1
go: finding github.com/spf13/cobra v0.0.3
go: finding golang.org/x/sys v0.0.0-20180905080454-ebe1bf3edb33
go: downloading mynewt.apache.org/newtmgr v0.0.0-20190801165443-4ba7e39aca9f
go: downloading mynewt.apache.org/newt v0.0.0-20190723172213-9f7829a76fef
go: extracting mynewt.apache.org/newt v0.0.0-20190723172213-9f7829a76fef
go: downloading github.com/sirupsen/logrus v1.4.2
go: extracting github.com/sirupsen/logrus v1.4.2
go: downloading golang.org/x/sys v0.0.0-20190422165155-953cdadca894
go: extracting golang.org/x/sys v0.0.0-20190422165155-953cdadca894
go: extracting mynewt.apache.org/newtmgr v0.0.0-20190801165443-4ba7e39aca9f
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/go-ble/ble v0.0.0-20181002102605-e78417b510a3
go: downloading github.com/cheggaaa/pb v1.0.18
go: downloading github.com/spf13/cast v1.3.0
go: downloading github.com/runtimeco/go-coap v0.0.0-20180109101419-47c23e2a63ea
go: downloading github.com/tarm/serial v0.0.0-20180830185346-98f6abe2eb07
go: downloading github.com/spf13/cobra v0.0.5
go: extracting github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/fatih/structs v1.0.0
go: extracting github.com/cheggaaa/pb v1.0.18
go: extracting github.com/spf13/cast v1.3.0
go: extracting github.com/tarm/serial v0.0.0-20180830185346-98f6abe2eb07
go: downloading golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3
go: downloading github.com/ugorji/go v1.1.7
go: downloading gopkg.in/abiosoft/ishell.v1 v1.0.0-20171224170712-50251d04cb42
go: extracting github.com/fatih/structs v1.0.0
go: downloading github.com/joaojeronimo/go-crc16 v0.0.0-20140729130949-59bd0194935e
go: extracting github.com/runtimeco/go-coap v0.0.0-20180109101419-47c23e2a63ea
go: downloading github.com/mattn/go-runewidth v0.0.2
go: extracting github.com/go-ble/ble v0.0.0-20181002102605-e78417b510a3
go: extracting gopkg.in/abiosoft/ishell.v1 v1.0.0-20171224170712-50251d04cb42
go: downloading github.com/pkg/errors v0.8.1
go: downloading github.com/mgutz/logxi v0.0.0-20161027140823-aebf8a7d67ab
go: extracting github.com/joaojeronimo/go-crc16 v0.0.0-20140729130949-59bd0194935e
go: downloading github.com/fatih/color v1.5.0
go: downloading github.com/flynn-archive/go-shlex v0.0.0-20150515145356-3f9db97f8568
go: downloading github.com/chzyer/readline v0.0.0-20171208011716-f6d7a1f6fbf3
go: extracting github.com/spf13/cobra v0.0.5
go: downloading github.com/spf13/pflag v1.0.3
go: extracting github.com/mattn/go-runewidth v0.0.2
go: extracting github.com/ugorji/go v1.1.7
go: downloading github.com/ugorji/go/codec v1.1.7
go: extracting github.com/pkg/errors v0.8.1
go: extracting github.com/mgutz/logxi v0.0.0-20161027140823-aebf8a7d67ab
go: downloading github.com/mattn/go-colorable v0.0.9
go: downloading github.com/mattn/go-isatty v0.0.2
go: extracting github.com/flynn-archive/go-shlex v0.0.0-20150515145356-3f9db97f8568
go: downloading github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
go: extracting github.com/chzyer/readline v0.0.0-20171208011716-f6d7a1f6fbf3
go: extracting github.com/spf13/pflag v1.0.3
go: extracting github.com/mattn/go-colorable v0.0.9
go: extracting github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
go: extracting github.com/mattn/go-isatty v0.0.2
go: extracting github.com/fatih/color v1.5.0
go: extracting github.com/ugorji/go/codec v1.1.7
go: extracting golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3
mynewt.apache.org/newtmgr/nmxact/lora
github.com/go-ble/ble
mynewt.apache.org/newt/util
github.com/go-ble/ble/linux/hci/socket
github.com/chzyer/readline
golang.org/x/net/context
mynewt.apache.org/newtmgr/nmxact/bledefs
github.com/go-ble/ble/linux/att
mynewt.apache.org/newtmgr/nmxact/nmxutil
mynewt.apache.org/newt/util/unixchild
mynewt.apache.org/newtmgr/nmxact/task
mynewt.apache.org/newtmgr/nmxact/nmcoap
mynewt.apache.org/newtmgr/nmxact/nmp
gopkg.in/abiosoft/ishell.v1
github.com/mitchellh/go-homedir
github.com/tarm/serial
github.com/go-ble/ble/linux/hci
mynewt.apache.org/newtmgr/newtmgr/core
mynewt.apache.org/newtmgr/nmxact/sesn
mynewt.apache.org/newtmgr/nmxact/omp
mynewt.apache.org/newtmgr/newtmgr/nmutil
mynewt.apache.org/newtmgr/nmxact/xport
mynewt.apache.org/newtmgr/nmxact/mgmt
github.com/go-ble/ble/linux
mynewt.apache.org/newtmgr/nmxact/mtech_lora
mynewt.apache.org/newtmgr/nmxact/nmserial
mynewt.apache.org/newtmgr/nmxact/nmble
mynewt.apache.org/newtmgr/nmxact/udp
mynewt.apache.org/newtmgr/nmxact/xact
mynewt.apache.org/newtmgr/newtmgr/bll
mynewt.apache.org/newtmgr/newtmgr/config
mynewt.apache.org/newtmgr/newtmgr/cli
github.com/apache/mynewt-mcumgr-cli/mcumgr
```

**Comment by ccollins476ad** [comment]
Thanks @SIGSTACKFAULT.  I opened a similar PR here: https://github.com/apache/mynewt-newtmgr/pull/137 (this uses the latest ishell rather than the old one).  Could those experiencing this problem please try my PR?  Unfortunately, I have no idea why I don't see this same build error, so I can't verify it myself.

If my PR doesn't work, then I agree we should merge the original.

EDIT: To clarify, the only reason I opened a separate PR is the language on the ishell page ("Kindly change your import path to...").  @SIGSTACKFAULT, if you want to change your PR to use the latest ishell, that is fine too!

**Comment by utzig** [comment]
I was taking a look at the `ishell` docs because someone asked me in pvt if we were not dependent on `v1`. `v1` uses a different API with `Register` and `Start` methods, while we use a `Context` "object" and `AddCmd` from v2. Also from the documentation it seem like it would be better to settle on `v2`, so I will close #136 which works right now but might break in the future and just reopened #137.

**Comment by utzig** [comment]
The issue on `newtmgr` was fixed by apache/mynewt-newtmgr#137, but this repo still needs update to fetch the latest `newtmgr`...

**Comment by utzig** [comment]
@matwiszynmarek @lawrencek52 @SIGSTACKFAULT This should be fixed now, thanks for the tips, and feel free to reopen if you still have issues.
